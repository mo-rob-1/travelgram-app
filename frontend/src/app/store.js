import { configureStore } from "@reduxjs/toolkit";
import authReducer from "../features/auth/authSlice";
import imageReducer from "../features/images/imageSlice";

export const store = configureStore({
  reducer: {
    auth: authReducer,
    images: imageReducer,
  },
});
