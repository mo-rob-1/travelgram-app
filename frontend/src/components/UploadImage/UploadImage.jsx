import { useState } from "react";
import { useDispatch } from "react-redux";
import { uploadImage } from "../../features/images/imageSlice";

function UploadImage() {
  const [image, setImage] = useState(null);

  const dispatch = useDispatch();

  const handleChange = (e) => {
    setImage(e.target.files[0]);
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    const formData = new FormData();
    formData.append("avatar", image);
    dispatch(uploadImage(formData));
  };

  return (
    <div>
      <form onSubmit={handleSubmit}>
        <input type="file" onChange={handleChange} />
        <button type="submit">Upload</button>
      </form>
    </div>
  );
}

export default UploadImage;
