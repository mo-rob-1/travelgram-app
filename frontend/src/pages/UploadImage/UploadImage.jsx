import { useState, useEffect } from "react";
import { useDispatch } from "react-redux";
import { uploadImage } from "../../features/images/imageSlice";
import { useNavigate } from "react-router-dom";
import { toast } from "react-toastify";
import {
  Section,
  Canvas,
  PreviewImage,
  Form,
  InputWrapper,
  Input,
  ButtonWrapper,
  Button,
  Wrapper,
  Title,
  ColOne,
  ColTwo,
  UploadInput,
} from "./UploadImage.styled";

function UploadImage() {
  const [image, setImage] = useState(null);
  const [caption, setCaption] = useState("");
  const [imageLocation, setImageLocation] = useState("");
  const [selectedFile, setSelectedFile] = useState();
  const [preview, setPreview] = useState();

  useEffect(() => {
    if (!selectedFile) {
      setPreview(undefined);
      return;
    }

    const objectUrl = URL.createObjectURL(selectedFile);
    setPreview(objectUrl);

    return () => URL.revokeObjectURL(objectUrl);
  }, [selectedFile]);

  const dispatch = useDispatch();
  const navigate = useNavigate();

  const handleChange = (e) => {
    setImage(e.target.files[0]);

    if (!e.target.files || e.target.files.length === 0) {
      setSelectedFile(undefined);
      return;
    }

    setSelectedFile(e.target.files[0]);
  };

  const cancel = () => {
    setImage(null);
    setSelectedFile(null);
    setCaption("");
    setImageLocation("");
  };

  const handleSubmit = (e) => {
    e.preventDefault();

    if (image === null) {
      toast.error("Please upload an image");
      return;
    } else {
      const formData = new FormData();
      formData.append("image", image);
      formData.append("caption", caption);
      formData.append("imageLocation", imageLocation);

      dispatch(uploadImage(formData));

      navigate("/");
    }
  };

  return (
    <Section>
      <Title>Upload Image</Title>
      <Form onSubmit={handleSubmit} data-testid="form">
        <ColOne>
          <div>
            <UploadInput
              type="file"
              onChange={handleChange}
              data-testid="image-upload-input"
              style={{ marginBottom: "1rem" }}
            />
          </div>

          <Wrapper>
            <InputWrapper>
              <label htmlFor="caption">Caption:</label>
              <br></br>
              <Input
                type="text"
                required
                value={caption}
                onChange={(e) => setCaption(e.target.value)}
                data-testid="caption-input"
              />
            </InputWrapper>
            <InputWrapper>
              <label htmlFor="image-location">Image Location:</label>
              <br></br>
              <Input
                type="text"
                required
                value={imageLocation}
                onChange={(e) => setImageLocation(e.target.value)}
                data-testid="image-location-input"
              />
            </InputWrapper>
          </Wrapper>
        </ColOne>

        <ColTwo>
          {!selectedFile ? <Canvas /> : <PreviewImage src={preview} alt="preview" data-testid="preview" />}

          <ButtonWrapper>
            <Button onClick={cancel}>Cancel</Button>
            <Button type="submit" data-testid="upload-button">
              Upload
            </Button>
          </ButtonWrapper>
        </ColTwo>
      </Form>
    </Section>
  );
}

export default UploadImage;
