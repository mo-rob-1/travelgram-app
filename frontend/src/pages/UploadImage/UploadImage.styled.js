import styled from "styled-components";

export const Section = styled.section`
  margin-top: 1rem;

  @media (min-width: 1024px) {
    margin-top: 3rem;
  }
`;

export const Canvas = styled.canvas`
  height: 300px;
  width: 100%;
  border: 2px solid lightgray;
  @media (min-width: 1440px) {
    height: 400px;
  }
`;

export const PreviewImage = styled.img`
  width: 100%;
`;

export const Form = styled.form`
  display: grid;
  gap: 1rem;
  @media (min-width: 768px) {
    grid-template-columns: repeat(2, 1fr);
    grid-template-rows: 1fr;
    gap: 3rem;
    margin-top: 3.5rem;
  }
  @media (min-width: 1440px) {
    gap: 5rem;
  }
`;

export const InputWrapper = styled.div`
  margin-bottom: 1rem;
`;

export const Input = styled.input`
  width: 100%;
  border: 1px solid lightgray;
  padding: 0.8rem 0.8rem;
  border-radius: 0.4rem;
  margin-top: 0.4rem;
`;

export const ButtonWrapper = styled.div`
  display: grid;
  gap: 1rem;
  margin-top: 0.8rem;
  @media (min-width: 768px) {
    grid-template-columns: repeat(2, 1fr);
    grid-template-rows: 1fr;
  }
`;

export const Button = styled.button`
  background-color: #222;
  color: #fff;
  border: none;
  padding: 0.8rem 0.8rem;
  cursor: pointer;
`;

export const Wrapper = styled.div`
  @media (min-width: 768px) {
    margin-top: 1rem;
  }
`;

export const Title = styled.h1`
  margin-top: 0;
  @media (min-width: 768px) {
    margin-top: 1rem;
  }
`;

export const ColOne = styled.div`
  display: block;
`;

export const ColTwo = styled.div`
  display: block;
`;
