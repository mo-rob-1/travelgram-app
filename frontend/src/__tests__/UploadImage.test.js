import { render } from "@testing-library/react";
import { createMemoryHistory } from "history";
import { Provider } from "react-redux";
import { store } from "../app/store";
import { BrowserRouter as Router } from "react-router-dom";
import UploadImage from "../pages/UploadImage/UploadImage";

describe("UploadImage", () => {
  it("renders the upload image page", () => {
    const history = createMemoryHistory();
    const { asFragment } = render(
      <Provider store={store} history={history}>
        <Router>
          <UploadImage />
        </Router>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders the form to upload image", () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <Router>
          <UploadImage />
        </Router>
      </Provider>
    );
    const form = getByTestId("form");
    expect(form).toBeTruthy();
  });

  it("renders the input to upload image", () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <Router>
          <UploadImage />
        </Router>
      </Provider>
    );
    const input = getByTestId("image-upload-input");
    expect(input).toBeTruthy();
  });

  it("renders the caption input", () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <Router>
          <UploadImage />
        </Router>
      </Provider>
    );
    const input = getByTestId("caption-input");
    expect(input).toBeTruthy();
  });

  it("renders the image location input", () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <Router>
          <UploadImage />
        </Router>
      </Provider>
    );
    const input = getByTestId("image-location-input");
    expect(input).toBeTruthy();
  });

  it("renders the upload button", () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <Router>
          <UploadImage />
        </Router>
      </Provider>
    );
    const button = getByTestId("upload-button");
    expect(button).toBeTruthy();
  });
});
