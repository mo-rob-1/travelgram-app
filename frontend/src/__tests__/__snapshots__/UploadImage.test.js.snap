// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UploadImage renders the upload image page 1`] = `
<DocumentFragment>
  <section
    class="sc-bczRLJ boIiCX"
  >
    <h1
      class="sc-papXJ bPCgLm"
    >
      Upload Image
    </h1>
    <form
      class="sc-hKMtZM kMwlze"
      data-testid="form"
    >
      <div
        class="sc-jqUVSM hyXLld"
      >
        <div>
          <input
            class="sc-iqcoie dmEgGu"
            data-testid="image-upload-input"
            style="margin-bottom: 1rem;"
            type="file"
          />
        </div>
        <div
          class="sc-ftvSup klfYgA"
        >
          <div
            class="sc-eCYdqJ XwHSn"
          >
            <label
              for="image-location"
            >
              Location:
            </label>
            <br />
            <input
              class="sc-jSMfEi ZyOou"
              data-testid="image-location-input"
              required=""
              type="text"
              value=""
            />
          </div>
          <div
            class="sc-eCYdqJ XwHSn"
          >
            <label
              for="caption"
            >
              Caption:
            </label>
            <br />
            <input
              class="sc-jSMfEi ZyOou"
              data-testid="caption-input"
              required=""
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
      <div
        class="sc-kDDrLX fZDAPW"
      >
        <canvas
          class="sc-gsnTZi lkApQR"
        />
        <div
          class="sc-gKXOVf ktAsrd"
        >
          <button
            class="sc-iBkjds dQtZOM"
          >
            Cancel
          </button>
          <button
            class="sc-iBkjds dQtZOM"
            data-testid="upload-button"
            type="submit"
          >
            Upload
          </button>
        </div>
      </div>
    </form>
  </section>
</DocumentFragment>
`;
